src/app/zap-templates/common/ClustersHelper.js

```diff
@@ -462,6 +462,48 @@ Clusters.init = function(context, packageId) {
   return Promise.all(promises).then(([types, clusters, commands, attributes, globalAttributes]) => {
     this._clusters = clusters;
     this._commands = enhancedCommands(commands, types);
+    // Work around ZAP's not-great support for mandatory global attributes by
+    // synthesizing them here.
+    // TODO: This should happen at the ZAP DB construction stage.
+    clusters.map(cluster => {
+      // Add AttributeList attribute.
+      const attrDef = {
+        // id should be unused, I think
+        id: "DO NOT USE THIS",
+        clusterId: cluster.id,
+        code: 0xFFFB,
+        manufacturerCode: 0,
+        hexCode: '0xFFFB',
+        name: "AttributeList",
+        side: "server",
+        type: "array",
+        entryType: "attrib_id",
+        minLength: 0,
+        maxLength: undefined,
+        min: undefined,
+        max: undefined,
+        storage: "External",
+        isIncluded: true,
+        isSingleton: false,
+        isBound: false,
+        isWritable: false,
+        isNullable: false,
+        mustUseTimedWrite: false,
+        defaultValue: undefined,
+        includedReportable: true,
+        minInterval: undefined,
+        maxInterval: undefined,
+        reportableChange: undefined,
+        define: "DO NOT USE THIS",
+      };
+      let index = attributes.findIndex(a => a.code > 0xFFFB);
+      if (index != -1) {
+        attributes.splice(index, 0, attrDef);
+      } else {
+        attributes.push(attrDef);
+      }
+    });
+    globalAttributes.push(0xFFFB);
```

Comment:
Mind adding another function called `addGlobalAttributeList` that modifies `globalAttributes` instead of sticking it here ? I suspect the same approach will be done for `commands`for both sides as well as events...

Reply:
This code is gone now that I added the `<globalAttribute>` in the XML.